internal/querycoordv2/meta/replica_manager.go

```diff
@@ -283,20 +271,88 @@ func (m *ReplicaManager) GetByResourceGroup(rgName string) []*Replica {
 	return ret
 }
 
-func (m *ReplicaManager) AddNode(replicaID typeutil.UniqueID, nodes ...typeutil.UniqueID) error {
+// RecoverNodesInCollection recovers all nodes in collection with latest resource group.
+// Promise a node will be only assigned to one replica in same collection at same time.
+// 1. Move the nodes to outbound if they are not in related resource group.
+// 2. Add new incoming nodes into the replica if they are not in-used by other replicas of same collection.
+// 3. replicas in same resource group will shared the nodes in resource group fairly.
+func (m *ReplicaManager) RecoverNodesInCollection(collectionID typeutil.UniqueID, rgs map[string]typeutil.UniqueSet) error {
+	if err := m.validateResourceGroups(rgs); err != nil {
+		return err
+	}
+
 	m.rwmutex.Lock()
 	defer m.rwmutex.Unlock()
 
-	replica, ok := m.replicas[replicaID]
+	// create a helper to do the recover.
+	helper, err := m.getReplicaAssignmentHelper(collectionID, rgs)
+	if err != nil {
+		return err
+	}
+
+	modifiedReplicas := make([]*Replica, 0)
+	// recover node by resource group.
+	helper.RangeOverResourceGroup(func(replicaHelper *replicaAssignmentHelper) {
+		replicaHelper.RangeOverReplicas(func(assignment *replicaAssignmentInfo) {
+			outboundNodes := assignment.GetNewOutboundNodes()
+			recoverableNodes, incomingNodeCount := assignment.GetRecoverNodesAndIncomingNodeCount()
+			// There may be not enough incoming nodes for current replica,
+			// Even we filtering the nodes that are used by other replica of same collection in other resource group,
+			// current replica's expected node may be still used by other replica of same collection in same resource group.
+			incomingNode := replicaHelper.AllocateIncomingNodes(incomingNodeCount)
```

Comment:
here will cause node num unbalance between replicas, maybe only few replicas could get enough nodes.

Reply:
We have calculated expectedNodes of every replicas (excepted outbound node in other replica of different resource group) in same resource group at construction of `newCollectionAssignmentHelper`.
So every replica will get balanced of current available node count.

However, when we need to balance the node count between different replica in same resource group.
We need to transfer some Read-Write nodes in one replica into Read only nodes, there will be a period of time unbalanced which cannot be avoid.